packages/@ember/-internals/routing/lib/services/router.ts

```diff
@@ -324,21 +323,31 @@ class RouterService<R extends Route> extends Service.extend(Evented) {
     let hasQueryParams = Object.keys(queryParams).length > 0;
 
     if (hasQueryParams) {
+      // UNSAFE: casting `routeName as string` here encodes the existing
+      // assumption but may be wrong: `extractRouteArgs` correctly returns it
+      // as `string | undefined`. There may be bugs if `_prepareQueryParams`
+      // does not correctly account for `undefined` values for `routeName`.
+      //  Spoilers: under the hood this currently uses router.js APIs which
+      // *do not* account for this being `undefined`.
+      let targetRouteName = routeName as string;
+
       queryParams = Object.assign({}, queryParams);
       this._router._prepareQueryParams(
-        // UNSAFE: casting `routeName as string` here encodes the existing
-        // assumption but may be wrong: `extractRouteArgs` correctly returns it
-        // as `string | undefined`. There may be bugs if `_prepareQueryParams`
-        // does not correctly account for `undefined` values for `routeName`.
-        //  Spoilers: under the hood this currently uses router.js APIs which
-        // *do not* account for this being `undefined`.
-        routeName as string,
+        targetRouteName,
```

Comment:
what happens when targetRouteName here is undefined?

should we remove the `as` and instead narrow the type with an assert?

Reply:
I don’t know. It was like this before and I just moved things around. I do want to review all `as` casts someone soon.